### ISSUE TITLE
[Feature request] Beta and Nightly Updates

### ISSUE THREAD

**Issue by alvis05** [webmin/webmin#968]
Can you please add a nightly and/or beta update option with in Webmin.

**Comment by rostovtsev** [comment]
Hi, we already have beta (development) channel.

http://www.webmin.com/devel.html

**Comment by alvis05** [comment]
@iliaross
Is there a way to automatically update to the development channel, instead of the stable channel.

**Comment by rostovtsev** [comment]
Yeah, just alter you package manager's repo URL to:

https://download.webmin.com/devel/rpm

or any other, based on your distro.

**Comment by alvis05** [comment]
@iliaross 
Do you mean
deb https://download.webmin.com/devel/deb/ sarge contrib

**Comment by swelljoe** [comment]
I don't think those are repos for yum or apt. They don't have metadata, I don't think. So, it'd need some changes on the backend to add metadata generation for those directories (which is very easy for yum, but stupidly tricky for apt). I don't have access to the master Webmin download server to do that, but it might be something Jamie can tackle when he's got more time.

**Comment by rebekah65** [comment]
@swelljoe 
A working Linux nightly repo would be great but, even an update script thats run via a cron would also be good or perhaps just making the update-from-repo.sh more cron frendly could also be good.

**Comment by rebekah65** [comment]
@iliaross 
Is there a way to stop it from reinstalling the same vershon if there's no new one available.

**Comment by rostovtsev** [comment]
I don't think so. But it's possible to wrap update script, and make a check in your own `.sh script`. I'll let Kay know. It's easily fixable.

**Comment by gnadelwartz** [comment]
I'll have a look how to implement to update only if a newer version is availible.
Currently it's not implemented because it can also used to downgrade or reinstall.

Question: How to select a newer Version? currently you can install stable versions or dev versions based on commits. Since there are daily commits, at least every day a new Version exist.

personally I would not use it from cron for productive environment, because you'll never know if something breaks ...

**Comment by rostovtsev** [comment]
@gnadelwartz Just by comparing strings in current version and in remote after the check.

I would do this optional, with the switch like `--soft` or something. :smile:

**Comment by rostovtsev** [comment]
I have started to think that it's not fair that Webmin and Usermin are the only ones that don't have built-in notification for new development version?

Should I add it, what do you think @jcameron and @swelljoe ? It would look like this and would give a notification in case enabled, on the Dashboard and into Notification panel any time there is new update available?

![screenshot from 2018-10-09 23-07-12](https://user-images.githubusercontent.com/4426533/46695551-8b1cc500-cc18-11e8-8ed5-a045e8ba1e08.png)

I will also try to provide a user with one click update, just like for the theme.

**Comment by rostovtsev** [comment]
@gnadelwartz Could you add a very simple option, that would prevent Webmin from restarting upon running force update?

When you run it from UI, Webmin should be restarted different way, in order not to interrupt the current process (self-update).

For example, something like:
```
./update-from-repo.sh -force -norestart
```

**Comment by swelljoe** [comment]
I'd be hesitant to make git versions visible to regular users. It should be a little hard to end up with something that could be broken in catastrophic ways (probably not, but sometimes things go wrong and commits have to be rolled back or fixed).

Making it easy for people who know they want it is one thing (so a command line tool to update easily is great!), making it easy for people to shoot themselves in the foot because they think the new one is always better is maybe dangerous?

But, *if* the packaged devel versions were to get repo support, then I'd be OK with it, because at least those versions have had Jamie's automated tests and a little bit of sanity checking before being packaged up. So...yes to GUI support for devel package installation, no to GUI support for git versions.

**Comment by swelljoe** [comment]
Then again, if they become repos, we don't need GUI support. The user just adds the repo to their yum or apt configuration, and they'll get the devel updates. 

So: No to GUI support for it. It's unneeded for devel versions if we get repo metadata added, and git versions are too risky for non-technical users to accidentally end up with.

**Comment by rebekah65** [comment]
@iliaross That looks really good although, based on what @swelljoe said it may be worth adding a popup warning message. That tells users that this version is unstable and may cause problems and need debugging, and require them to click agree be for installing.

**Comment by rostovtsev** [comment]
@swelljoe Joe, do you think having an explicit warning wouldn't be safe enough?

Maybe create a separate section for it? Like Developer Zone and add there all unstable update toggles, with notification and etc.?

**Comment by rostovtsev** [comment]
Okay, I think I will put off this idea for the future, and add this to the list of possible (not must have) improvements.

It's not trivial to run updates from within Webmin, while changing its core files.

The theme configuration option for showing theme beta updates was moved to `authentic-theme/config` file and must be controlled from over there.
